keps/sig-instrumentation/4828-component-flagz/README.md

```diff
@@ -0,0 +1,388 @@
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+We propose to enhance the observability, troubleshooting, and debugging capabilities of core Kubernetes components by implementing a flagz endpoint for all components. This will provide users with real-time visibility into the command-line flags each component was started with, aiding in the diagnosis of configuration issues and unexpected behavior
+
+## Motivation
+
+This proposal extends the ideas presented in the [Component Statusz KEP](https://github.com/kubernetes/enhancements/pull/4830), however, it proposes a distinct flagz endpoint to enhance visibility into component configurations. By providing real-time visibility into the active configuration of Kubernetes components, flagz enables users to quickly spot misconfigurations that could lead to system instability or outages
+
+### Goals
+
+* Provide users with the ability to dynamically inspect and understand the active flags of running Kubernetes components
+
+### Non-Goals
+
+* Replace existing monitoring solutions: metrics, logs, traces
+* Provide information about components that are not directly accessible due to network restrictions
+
+## Proposal
+
+We are proposing to add a new endpoint, /flagz on all core Kubernetes components which returns the command-line flags each component was started with.
+
+### Risks and Mitigations
+
+1. **No sensitive data exposed**
+    
+    We will ensure that no sensitive data is exposed through flagz and that access to this endpoint is gated by using system-monitoring group.
```

Comment:
how?  Do flag definitions allow us to express the idea that the flag is "sensitive" ?

Reply:
I dont think there currently is support for marking flags like that. So we would need to introduce a way to express this. 
cc @cjcullen who I just reached out to for his thoughts on the same. Also @liggitt

Reply:
Do we need to push that up to pflag?  One could argue that falgs should NEVER have sensitive information, since they are visible via `ps`.

Reply:
> Do we need to push that up to pflag? 

Or.. maybe we could create a separate set of "sensitive flags" and while printing out the values in the [flagz handler](https://github.com/kubernetes/kubernetes/pull/127581/files#r1791055661), redact the values for those flags? 

> One could argue that falgs should NEVER have sensitive information, since they are visible via ps

That's indeed true. But if we still feel that exposing this info in an endpoint will be easier for attackers to get hold of, we can impose some restrictions on what data to expose in the endpoint.

Reply:
Do we have any examples of these sorts of flags?

Reply:
Perhaps the [TLS](https://github.com/richabanker/kubernetes/blob/master/cmd/kube-apiserver/app/options/options.go#L146-L150) related flags.. (Also waiting for someone from sig-auth to help answer that..)

Reply:
I'm not aware any flags in long-lived components (kube-apiserver, kubelet, scheduler, controller-manager, etc) that contain sensitive information directly. For things like TLS or credentials, they always point at files which contain the data.

There _are_ shorter-lived invocations that take flags containing sensitive information (like `kubeadm init` with `--certificate-key` or `--token`), and components built that bind client-go user flags (like `kubectl`) *can* specify a token credential on the command-line using `--token`. Both of those have the option to consume those values from a file as well.

Reply:
Great! Thanks for pointing out the pflag annotation feature.  That's perfect for displaying the non-CLASSIFIED flags.